packages/material-ui/src/OverridableComponent.d.ts

```diff
@@ -0,0 +1,43 @@
+import * as React from 'react';
+import { Omit } from '.';
+import { StyledComponentProps } from './styles';
+
+export interface OverridableComponent<M extends OverridableTypeMap> {
+  <C extends React.ReactType>(props: { component: C } & OverrideProps<M, C>): JSX.Element;
+  (props: DefaultProps<M>): JSX.Element;
+}
+
+// prettier-ignore
+export type OverrideProps<
+  M extends OverridableTypeMap,
+  C extends React.ReactType
+> = (
+  & BaseProps<M>
+  & Omit<React.ComponentPropsWithRef<C>, keyof UniversalProps<M>>
+);
+
+// prettier-ignore
+export type DefaultProps<M extends OverridableTypeMap> =
+  & BaseProps<M>
+  & Omit<React.ComponentPropsWithRef<M['defaultComponent']>, keyof BaseProps<M>>;
+
+// prettier-ignore
+export type BaseProps<M extends OverridableTypeMap> =
+  & M['props']
+  & UniversalProps<M>;
+
+export interface UniversalProps<M extends OverridableTypeMap>
+  extends StyledComponentProps<M['classKey']> {
+  className?: string;
+  style?: React.CSSProperties;
+}
+
+export interface OverridableTypeMap {
+  props: {};
+  defaultComponent: React.ReactType;
+  classKey: string;
+}
+
+export type Simplify<T> = { [K in keyof T]: T[K] };
```

Comment:
We can tackle this later when we encounter union types (e.g. `TextField`)

Reply:
>Is it fair to say that this only squashes intersection the originated from e.g. `P & { children?: ReactNode }` inferrence?

I honestly don't remember the rationale for this, lol. But I see that making it

```ts
export type Simplify<T> = T
```

seems to break things in mysterious ways 🤷‍♂️

>Is it maybe better to use [this neat trick here](https://github.com/Microsoft/TypeScript/issues/28483#issuecomment-438084079) so that it doesn't "destroy" union types from e.g. `TextField`?

Yes, it seems like a good idea to allow for unions. I'll go ahead and make it

```ts
export type Simplify<T> = T extends any ? { [K in keyof T]: T[K] } : never;
```